[lldb][swift] Filter await-resume funclets when setting breakpoints

felipepiovezan opened the pull request:
These funclets only serve to `task_dealloc` previously allocated tasks when returning from an async call, and immediately `task_switch` to the next await-suspend funclet (which contains real user code).

By not filtering out these funclets, any breakpoint on a line with an async call will cause execution to pause 3 times: once before the call, twice when "returning" from the call, which makes for a confusing experience.

The patch does the filtering on `BreakpointResolver::SetSCMatchesByLine`, which is the common code between BreakpointResolverFileLine and BreakpointResolverFileRegex.

We also considered changing the debug line information in any of the many different lowering stages the swift compiler, but this turned out to be very complex to do in a targeted way; more often than not, a handful of early-IR coroutine instructions get expanded into multiple function clones, all inheriting the same debug line information. The current approach also has the advantaged of being easily reversible if we decide to do so.

felipepiovezan commented:
I'm very open to other ways of accomplishing this, in case anyone has suggestions

jimingham commented:
Filtering them in the File & Line resolver seems fine to me.

However, there needs to be a `virtual Language::RemoveInvalidSourceLineMatches` or something like that which does nothing.  Then your function needs to implement that in the Swift Language.  I'm not wedded to the name, but you have to make this generic.  We really want to keep from adding more new #ifdef LLDB_ENABLE_SWIFT in generic files.

jimingham commented:
You should probably also do the generic stuff on the llvm.org version, cherry-pick that and add your implementation so we don't end up with more diffs on the swift side.

felipepiovezan commented:
> You should probably also do the generic stuff on the llvm.org version, cherry-pick that and add your implementation so we don't end up with more diffs on the swift side.

I actually had that on my first iteration of this patch, but then realized we would be putting virtual methods with no real implementation upstream, which is generally something that sees push back against (it would basically be dead code, since Swift is not an upstreamed language). If you believe this won't be a problem, I can go the language route (still need to figure out if we have access to a `Language` object from this point)

jimingham commented:
This is a general feature, where the compiler is emitting some debug information that for its own internal reasons it wishes it could manage not to emit, but can't.  OTOH, the debugger after the fact can.  There's nothing swift specific about that, except currently the swift compiler is the only one that does it.  So I don't think it will be an issue putting in an API here - and that's much preferable to more `#ifdef LLDB_ENABLE_SWIFT`.

The breakpoints offer an option to filter by language, so you should be able to get your hands on the language of the function you are setting the breakpoint in.

felipepiovezan committed:
[lldb][swift] Filter await-resume funclets when setting breakpoints

These funclets only serve to `task_dealloc` previously allocated tasks when
returning from an async call, and immediately `task_switch` to the next
await-suspend funclet (which contains real user code).

By not filtering out these funclets, any breakpoint on a line with an async call
will cause execution to pause 3 times: once before the call, twice when
"returning" from the call, which makes for a confusing experience.

The patch does the filtering on `BreakpointResolver::SetSCMatchesByLine`, which
is the common code between BreakpointResolverFileLine and
BreakpointResolverFileRegex.

We also considered changing the debug line information in any of the many
different lowering stages the swift compiler, but this turned out to be very
complex to do in a targeted way; more often than not, a handful of early-IR
coroutine instructions get expanded into multiple function clones, all
inheriting the same debug line information. The current approach also has the
advantaged of being easily reversible if we decide to do so.

felipepiovezan commented:
Sounds good, I will convert this to a draft while I propose the first commit of this PR upstream

felipepiovezan commented:
Upstream patch https://github.com/llvm/llvm-project/pull/83908

adrian-prantl commented on lldb/include/lldb/Target/Language.h:
```diff
@@ -337,6 +337,10 @@ class Language : public PluginInterface {
 
   virtual llvm::StringRef GetInstanceVariableName() { return {}; }
 
+  virtual bool IsInterestingCtxForLineBreakpoint(const SymbolContext &) const {
```
Doxygen comment?
Also we can probably spell out Ctx?

adrian-prantl commented on lldb/source/Breakpoint/BreakpointResolver.cpp:
```diff
@@ -200,13 +201,24 @@ bool operator<(const SourceLoc lhs, const SourceLoc rhs) {
     llvm::StringRef log_ident, uint32_t line, std::optional<uint16_t> column) {
   llvm::SmallVector<SymbolContext, 16> all_scs;
   for (uint32_t i = 0; i < sc_list.GetSize(); ++i)
     all_scs.push_back(sc_list[i]);
```
Why not just apply the filter right here before entering it into the list?

adrian-prantl commented on lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp:
```diff
@@ -1793,6 +1793,17 @@ SwiftLanguage::GetDemangledFunctionNameWithoutArguments(Mangled mangled) const {
+  // In async functions, ignore "await resume" funclets, these only deallocate
+  // the async context and task_switch back to user code.
+  auto demangled = sc.function->GetMangled().GetDemangledName(&sc);
+  return demangled.GetStringRef().contains("await resume");
```
A much better way to implement this would be to walk the demangle tree.
Also this function will incorrectly return true if I had a function that takes an "await resume" function pointer as argument. By just checking the outermost nodes of the demangle tree you can address both problems.

Look at the output of `swift-demangle -expand <mangled name>` and copy any of the functions in TypeSystemSwiftTypeRef.cpp.

felipepiovezan replied:
> will incorrectly return true if I had a function that takes an "await resume" function pointer as argument

Is this possible? these functions are created by the corosplitter

adrian-prantl replied:
Probably not, but who knows what future runtime optimizations will be trying to do.

adrian-prantl replied:
In SwiftLanguageRuntimeNames.cpp there is IsSwiftAsyncFunctionSymbol which you might be able to adapt.

felipepiovezan replied:
I've looked into those and they seem like a nicer way to do these checks indeed.
That said, there are at least two bugs in their implementation that we'll need to fix before being able to reuse them.

felipepiovezan commented:
I've re-implemented the swift side of things with the suggestions and marked this patch as ready for review.

Note that the first commit here is identical to https://github.com/llvm/llvm-project/pull/83908/files, and I'll convert it to an actual cherry-pick once it gets merged upstream (won't merge this PR until that's done)

adrian-prantl commented on lldb/include/lldb/Target/Language.h:
```diff
@@ -337,6 +337,15 @@ class Language : public PluginInterface {
+  // by line (number or regex). This is useful for languages that create
+  // artificial functions without any meaningful user code associated with them
+  // (e.g. code that gets expanded in late compilation stages, like by
+  // CoroSplitter).
```
///

adrian-prantl commented on lldb/include/lldb/Target/Language.h:
```diff
@@ -337,6 +337,15 @@ class Language : public PluginInterface {
+  // artificial functions without any meaningful user code associated with them
+  // (e.g. code that gets expanded in late compilation stages, like by
+  // CoroSplitter).
+  virtual bool IsArtificialCtxForLineBreakpoint(const SymbolContext &) const {
```
how about `IgnoreForLineBreakpoints` ?

adrian-prantl commented on lldb/source/Breakpoint/BreakpointResolver.cpp:
```diff
@@ -204,8 +205,18 @@ void BreakpointResolver::SetSCMatchesByLine(
+          ->GetTarget()
+          .GetIgnoreBreakpointsFromLanguageArtificialLocations();
+
+  for (uint32_t i = 0; i < sc_list.GetSize(); ++i) {
```
can this be range-based?

felipepiovezan replied:
It can, I just didn't want to add to the diff. As much as github is saying this is new code, it's not new code...

adrian-prantl commented on lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp:
```diff
@@ -1794,6 +1794,17 @@ SwiftLanguage::GetDemangledFunctionNameWithoutArguments(Mangled mangled) const {
+  StringRef name = sc.function->GetMangled().GetMangledName().GetStringRef();
+  // In async functions, ignore "await resume" funclets, these only deallocate
+  // the async context and task_switch back to user code.
+  return !SwiftLanguageRuntime::IsSwiftAsyncAwaitResumePartialFunctionSymbol(name);
```
Great!

felipepiovezan commented:
@adrian-prantl I've addressed your comments on the upstream pr https://github.com/llvm/llvm-project/pull/83908

adrian-prantl commented on lldb/include/lldb/Target/Language.h:
```diff
@@ -337,6 +337,15 @@ class Language : public PluginInterface {
+  virtual bool IgnoreForLineBreakpoints(const SymbolContext &) const {
+    return false;
+  }
+
```
Side note: let's try to upstream this abstract interface later.

felipepiovezan committed:
[lldb] Allow languages to filter breakpoints set by line (#83908)

Some languages may create artificial functions that have no real user
code, even though there is line table information for them. One such
case is with coroutine code that receives the CoroSplitter
transformation in LLVM IR. That code transformation creates many
different Functions, cloning one Instruction into many Instructions in
many different Functions and copying the associated debug locations.

It would be difficult to make that pass delete debug locations of cloned
instructions in a language agnostic way (is it even possible?), but LLDB
can ignore certain locations by querying its Language APIs and having it
decide based on, for example, mangling information.

(cherry picked from commit 0adccd1a7fd8e30650f76bd33471de28b7939455)

felipepiovezan commented:
Now with the official cherry-pick of the upstream lldb commit.

felipepiovezan commented:
@swift-ci test

kastiglione commented on lldb/source/Target/TargetProperties.td:
```diff
@@ -360,3 +360,9 @@ let Definition = "thread" in {
 }
+
+let Definition = "language" in {
+  def EnableFilterForLineBreakpoints: Property<"enable-filter-for-line-breakpoints", "Boolean">,
```
Is there a naming convention around boolean properties? Should this be prefixed with enable-, or not? It seems possibly redundant to me. Maybe the name should be `filter-for-line-breakpoints` and the values `enable` and `disable`? I'm not strongly opposed to the name, but it did strike me as a little unconventional.

felipepiovezan replied:
Current practice has all "enable"-like properties using boolean values, and most of them mentioning "enable" in the actual setting flag, though there are exceptions to this second point

felipepiovezan commented:
Superseded by https://github.com/swiftlang/llvm-project/pull/9112
